examples/voting/src/lib.rs

```diff
@@ -0,0 +1,415 @@
+#[derive(Debug, Serialize, SchemaType)]
+pub struct VoteEvent {
+    /// The account that casts the vote.
+    voter:      AccountAddress,
+    /// The index of the voting option that the account is voting for.
+    vote_index: VoteIndex,
+}
+
+/// The event logged by this smart contract.
+#[derive(Debug, Serial, SchemaType)]
+pub enum Event {
+    /// The event is logged when a new (or replacement) vote is cast by an
+    /// account.
+    Vote(VoteEvent),
+}
+
+// Contract functions
+
+/// Initialize the contract instance and start the election.
+/// A description, the vector of all voting options, and an end_time
+/// have to be provided.
+#[init(contract = "voting", parameter = "InitParameter", event = "Event")]
+fn init<S: HasStateApi>(
+    ctx: &impl HasInitContext,
+    state_builder: &mut StateBuilder<S>,
+) -> InitResult<State<S>> {
+    // Parse the parameter.
+    let param: InitParameter = ctx.parameter_cursor().get()?;
+    // Calculate the number of voting options.
+    let num_options = param.options.len() as u32;
+
+    // Set the state.
+    Ok(State {
+        description: param.description,
+        ballots: state_builder.new_map(),
+        tally: state_builder.new_map(),
+        end_time: param.end_time,
+        options: param.options,
+        num_options,
+    })
+}
+
+/// Enables accounts to vote for a specific voting option. Each account can
+/// change its selected voting option with this function as often as it desires
+/// until the end_time is reached.
+///
+/// It rejects if:
+/// - It fails to parse the parameter.
+/// - A contract tries to vote.
+/// - It is past the end_time.
+#[receive(
+    contract = "voting",
+    name = "vote",
+    mutable,
+    enable_logger,
+    parameter = "VoteParameter",
+    error = "VotingError"
+)]
+fn vote<S: HasStateApi>(
+    ctx: &impl HasReceiveContext,
+    host: &mut impl HasHost<State<S>, StateApiType = S>,
+    logger: &mut impl HasLogger,
+) -> VotingResult<()> {
+    // Check that the election hasn't finished yet.
+    ensure!(ctx.metadata().slot_time() <= host.state().end_time, VotingError::VotingFinished);
+
+    // Ensure that the sender is an account.
+    let acc = match ctx.sender() {
+        Address::Account(acc) => acc,
+        Address::Contract(_) => return Err(VotingError::ContractVoter),
+    };
+
+    // Parse the parameter.
+    let param: VoteParameter = ctx.parameter_cursor().get()?;
+
+    let new_vote_index = param.vote_index;
+
+    // Check that vote is in range
+    ensure!(new_vote_index < host.state().num_options, VotingError::InvalidVoteIndex);
+
+    // Insert or replace the vote for the account.
+    host.state_mut()
+        .ballots
+        .entry(acc)
+        .and_modify(|vote| *vote = new_vote_index)
+        .or_insert(new_vote_index);
+
+    // Update the tally for the `new_vote_index` with one additional vote.
+    host.state_mut().tally.entry(new_vote_index).and_modify(|vote| *vote += 1).or_insert(1);
```

Comment:
Is there a specific reason for using this pattern as opposed to maybe a bit easier to read
```rust
*host.state_mut().tally.entry(new_vote_index).or_insert(0) += 1
```